### PR TITLE
Fix time issues

### PR DESCRIPTION
**Pull request by huitema**
Issue #1601 points out that using `gettimeofday()` for managing timers can cause errors if the local admin resets that time. It is preferable on Linux to use `clock_gettime(CLOCK_MONOTONIC)` for that purpose. This requires separating the TLS related usage of time, which has to track "wall time", from the timer-related usage prevalent in QUIC protocol handling.

The solution here is:

* make sure that ticket and token related calls use `picoquic_get_tls_time(quic)`, and rewrite the token and ticket API so as to *not* take a time parameter;
* incorporate the patch proposed by @fbussery and use `CLOCK_MONOTONIC` for other protocol actions.

### CODE REVIEW

**Comment by huitema** [comment]
Last point: clock_gettime may in some cases return an error, but merely reading the clock should only cause an issue if the calling arguments are malformed. I would rather ignore the return code than create an error branch that is never exercised.

File: picoquictest/tls_api_test.c [hunk 1]
```diff
@@ -5117,7 +5116,7 @@ int virtual_time_test()
         {
             /* Check that the simulated time follows the simulation */
             for (int i = 0; ret == 0 && i < 5; i++) {
-                simulated_time += 12345678;
+                simulated_time += 12345000;
```
Comment (fbussery) [inline picoquictest/tls_api_test.c:5119]: To have same behaviour as before, may be we should have simulated_time += 12345678000 ?

File: picoquictest/ticket_store_test.c [hunk 2]
```diff
@@ -118,20 +118,31 @@ int ticket_store_test()
...
+    }
+
+    if (ret == 0) {
+        /* Writing an empty file */
+        ret = picoquic_save_tickets(p_first_ticket, current_time, test_ticket_file_name);
+    }
 
     /* Load the empty file again */
     if (ret == 0) {
-        ret = picoquic_load_tickets(&p_first_ticket_empty, retrieve_time, test_ticket_file_name);
```
Comment (fbussery) [inline picoquictest/ticket_store_test.c:127]: p_first_ticket_empty still defined in line 111. Not needed any more

File: picoquictest/tls_api_test.c [hunk 3]
```diff
@@ -5116,7 +5116,7 @@ int virtual_time_test()
         {
             /* Check that the simulated time follows the simulation */
             for (int i = 0; ret == 0 && i < 5; i++) {
-                simulated_time += 12345000;
+                simulated_time += 12345678000;
```
Comment (fbussery) [inline picoquictest/tls_api_test.c:5119]: I have a doubt.  In my mind, simulated_time was of same unity as picoquic_get_quic_time which did not change

**Comment by huitema** [comment]
I also have a doubt. The "gettimeofday" approach has the nice property of inheriting clock synchronization, NTP, etc. This is not strictly required for the transport, but it does make debugging easier, e.g., when comparing time stamps at the client and the server. I wonder whether the use of monotonic clock should be controlled by a compile option.

**Pull request merged**